extensions/amp-link-rewriter/0.1/link-rewriter.js

```diff
@@ -0,0 +1,205 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getConfigOpts} from './config-options';
+import {getDataParamsFromAttributes} from '../../../src/dom';
+
+export const
+    CTX_ATTR_NAME = 'shiftedctx',
+    CTX_ATTR_VALUE = () => {
+      return Date.now();
+    },
+    WL_ANCHOR_ATTR = [
+      'href',
+      'id',
+      'rel',
+      'rev',
+    ],
+    PREFIX_DATA_ATTR = /^vars(.+)/;
+
+export class LinkRewriter {
+  /**
+   * @param {!AmpElement} ampElement
+   * @param {?../../../src/service/viewer-impl.Viewer} viewer
+   */
+  constructor(ampElement, viewer) {
+    /** @private {?../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = viewer;
+
+    /** @private {?Event} */
+    this.event_ = null;
+
+    /** @public {string} */
+    this.rewrittenUrl = '';
+
+    /** @private {?Object} */
+    this.configOpts_ = getConfigOpts(ampElement);
+
+    /** @private {boolean|number|string} */
+    this.ctxAttrValue_ = CTX_ATTR_VALUE().toString();
+
+    /** @private {?RegExp} */
+    this.regexDomainUrl_ = /^https?:\/\/(www\.)?([^\/:]*)(:\d+)?(\/.*)?$/;
+
+    /** @private {Promise<Object>} */
+    this.vars_ = this.viewer_.getReferrerUrl().then(referrerUrl => {
+      const pageAttributes = {
+        referrer: referrerUrl,
+        location: this.viewer_.getResolvedViewerUrl(),
```

Comment:
there are 3 types of URL in AMP

CDN URL: `https://amp-example-com.cdn.ampproject.org/c/s/amp.example.com/`
Source URL: `https://amp.example.com/`
Canonical URL: `https://www.example.com/`

your current implementation will give you the CDN URL. If that's not what you want, consider use `Services.documentInfoForDoc(ampDoc).sourceUrl|canonicalUrl`